### PR TITLE
Form with controls and hooks

### PR DESCRIPTION
**Pull request by Matushl**
## Decisions made:
- removed all actions but submit actions from form (compared to Wue)
- ~~added i18n with hardcoded translations - there will be need for another task to make messages customizable~~

## Consciously left out (todos for the future PRs):
- add ui-informed storybook deployment into CI
  - move storybook (and maybe others into root dependencies)
- add more stories for whole Form, etc
- ~~add possibility to customize translations~~ (_alredy in this PR_)
- setup jest and add tests
- add ArrayControl (here) + FieldSet (into utima/ui)
- ~~? create some general `FiledCol` as Form ui element~~ (_columns was removed completely_)

### CODE REVIEW

File: packages/ui-informed/src/hooks/useFormSubmitActions.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+type UseFormSubmitActionsParams<T> = {
+  onSubmit?: (formState: TypedFormState<T>) => Promise<unknown>;
+};
+
+/**
+ * Form component actions hook. Handles form submit, delete and failure
+ * actions definitions. These actions also handle confirmation modals
+ * and notification messages.
+ */
+export function useFormSubmitActions<T>({
```
Comment (jsimck) [inline packages/ui-informed/src/hooks/useFormSubmitActions.tsx:16]: can we rename it to  `useDefaultSubmitAction` so it's more explicit?

File: packages/ui-informed/src/hooks/useFormSubmitActions.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  const handleSubmit = useCallback(
+    async (formState: FormState) => {
+      setIsSubmitting(true);
+
+      try {
+        await onSubmit?.(formState as TypedFormState<T>);
+
+        // Show success notification
+        toast.success(t('actions.success.title'), {
+          description: t('actions.success.message'),
```
Comment (jsimck) [inline packages/ui-informed/src/hooks/useFormSubmitActions.tsx:35]: as mentioned below, pass this as props.

File: packages/ui-informed/src/controls/formControl/FormControl.tsx
Comment (jsimck) [inline packages/ui-informed/src/controls/formControl/FormControl.tsx:1]: can we move this to root level formControl directory? `src/formControl` and other controls to separate directories?

File: packages/ui-informed/src/hooks/useFormSubmitActions.tsx
Comment (jsimck) [inline packages/ui-informed/src/hooks/useFormSubmitActions.tsx:1]: I would move this to `form` directory as it's tightly coupled to that component

File: packages/ui-informed/src/hooks/useFormContext.ts
Comment (jsimck) [inline packages/ui-informed/src/hooks/useFormContext.ts:1]: I would move this to `form` directory as it's tightly coupled to that component

File: packages/ui-informed/src/form/Form.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import { useTranslation } from 'react-i18next';
+import type { ZodObject, ZodRawShape } from 'zod';
+
+import { FormContext, type FormContextType } from '../hooks/useFormContext';
+import {
+  useFormSubmitActions,
+  type TypedFormState,
+} from '../hooks/useFormSubmitActions';
+
+export type FormProps<T> = Omit<
```
Comment (jsimck) [inline packages/ui-informed/src/form/Form.tsx:13]: can you rewrite all intersections (`&`) to interfaces with extend as it is in `utima/ui` (it's slight performance improvement in TSC but counts in a long run)

File: packages/ui-informed/src/form/Form.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  disabled?: boolean;
+  readOnly?: boolean;
+  loading?: boolean;
+  columns?: boolean;
+  onSubmit?: (formState: TypedFormState<T>) => Promise<unknown>;
+  zodSchema?: ZodObject<ZodRawShape>;
+};
+
+/**
+ * Basic form component used to wrap form inputs. It provides action bar,
```
Comment (jsimck) [inline packages/ui-informed/src/form/Form.tsx:27]: there is no action bar and column layout in this case :)

File: packages/ui-informed/src/form/Form.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ * which handles form state and validation.
+ */
+export function Form<T>({
+  onSubmit,
+  children,
+  initialValues,
+  readOnly = false,
+  disabled = false,
+  loading = false,
+  columns = true,
```
Comment (jsimck) [inline packages/ui-informed/src/form/Form.tsx:38]: I would omit columns entirely, this is app specific, so can be done easily by wrapping this component in the application (or rendering columns in different components)

File: packages/ui-informed/src/form/Form.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    [initialValues, disabled, loading, isSubmitting, readOnly, zodSchema],
+  );
+
+  return (
+    <InformedForm
+      initialValues={initialValues as Record<string, unknown>}
+      onSubmit={handleSubmit}
+      onSubmitFailure={handleSubmitFailure}
+      errorMessage={{
+        required: t('errors.required'),
```
Comment (jsimck) [inline packages/ui-informed/src/form/Form.tsx:65]: pass translations with props

File: packages/ui-informed/src/controls/formControl/FormMessage.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { cn } from '@utima/ui';
+import { forwardRef, type ComponentProps } from 'react';
+
+export type FormMessageProps = ComponentProps<'p'>;
+
+export const FormMessage = forwardRef<HTMLDivElement, FormMessageProps>(
```
Comment (jsimck) [inline packages/ui-informed/src/controls/formControl/FormMessage.tsx:6]: replace forward ref with `memo`, it's not needed in this case (done incorrectly in wue)

File: packages/ui-informed/src/controls/formControl/FormInfo.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import { cn } from '@utima/ui';
+import { forwardRef, type ComponentProps } from 'react';
+
+export type FormInfoProps = ComponentProps<'span'>;
+
+export const FormInfo = forwardRef<HTMLSpanElement, FormInfoProps>(
```
Comment (jsimck) [inline packages/ui-informed/src/controls/formControl/FormInfo.tsx:6]: replace forward ref with `memo`, it's not needed in this case (done incorrectly in wue)

File: packages/ui-informed/src/i18n/cs/translation.json
[the remark this answers is not in the document]
Comment (jsimck) [reply packages/ui-informed/src/i18n/cs/translation.json:1]: thinking about it now... I would probably move all translations to root `Form` component and pass it down through context. In this case we only need to "wrap" the root component with initial settings.. Or maybe do a completely new context provider which provides these (defined globally like react query provider etc.)

File: packages/ui-informed/src/zodUtils.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  type ZodType,
+  ZodOptional,
+  ZodObject,
+} from 'zod';
+
+export function isZodError(error: unknown) {
+  return error instanceof ZodError;
+}
+
+export function zodResolver(schema?: ZodType) {
```
Comment (jsimck) [inline packages/ui-informed/src/zodUtils.ts:13]: we probably can even export these functions, they can be usefull :)

File: packages/ui-informed/src/index.ts [hunk 10]
```diff
@@ -1,4 +1,31 @@
+/**
+ * FormWithControls
+ */
+export * as FormWithControls from './form';
```
Comment (jsimck) [inline packages/ui-informed/src/index.ts:4]: I am on the edge with this one, I am leaning towards not defining these aliases in this case and simply having the application use `import * as Form from '@utima/ui-informed';`.

There would be cases like `Form.Form` but you could also define custom alias, or use it without one. WHat do you think?

File: packages/ui-informed/src/index.ts [hunk 11]
```diff
@@ -1,4 +1,31 @@
...
+ */
+export { FormControl } from './controls/formControl/FormControl';
+
 /**
  * Controls
  */
+export { CheckboxControl } from './controls/CheckboxControl';
 export { InputControl } from './controls/InputControl';
+export { SelectControl } from './controls/SelectControl';
+export { TextAreaControl } from './controls/TextAreaControl';
```
Comment (jsimck) [inline packages/ui-informed/src/index.ts:17]: I would probably export these as names without Suffix. Maybe even drop the `Control` suffix entirely through these codebase (except FormControl). As it is 100% defined by the package itself that these components are controlled by infromed.

**Review by jsimck — CHANGES_REQUESTED** [review]

File: packages/ui-informed/src/index.ts (hunk 10 above)
Comment (Matushl) [reply packages/ui-informed/src/index.ts:4]: I just don't like that if you let it only on the import in final application, then there will be thinks like `Form.useDefaultSubmitActions`. But I admit that it's maybe kinda unusual to create this named import inside of package, but on the other hand we are directly offering user of our package "the best way" how to use it and do not mess up importing Checkbox from UI insted of UI-informed :-D 

But I don't have strong opinion on that, I just stated my thoughts and leaving decision up to you (so far I'm not removing it)

File: packages/ui-informed/src/i18n/cs/translation.json
[the remark this answers is not in the document]
Comment (Matushl) [reply packages/ui-informed/src/i18n/cs/translation.json:1]: I've implemented FormTranslationsContext, it provides english translations as default and you can override them with wrapping your app with provider with value of your translations

**Review by jsimck — APPROVED** [review]

**Pull request merged**